### DISCUSSION TITLE
"Rebase and merge" should autosquash fixup commits

### DISCUSSION THREAD

**Discussion by dhimmel** [community/community#40804] [Pull Requests]
### Select Topic Area

Product Feedback

### Body

It would be a killer feature if the [**Rebase and merge**](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges#rebase-and-merge-your-commits) method of merging a pull request would apply the `git rebase --autosquash` [argument](https://git-scm.com/docs/git-rebase#Documentation/git-rebase.txt---no-autosquash):

> When the commit log message begins with "squash! ...​" or "fixup! ...​" or "amend! ...​", and there is already a commit in the todo list that matches the same `...`, automatically modify the todo list of `rebase -i`, so that the commit marked for squashing comes right after the commit to be modified, and change the action of the moved commit from `pick` to `squash` or `fixup` or `fixup -C` respectively.

This would allow users to add commits to a PR using `git commit --fixup`, where fixup is applied to another commit in the PR but not the base branch. Then when rebase merging via GitHub the fixup commits would be automatically squashed into the specified commit. It would save users from having to perform the `git rebase --autosquash --interactive` locally and having to force push to the PR branch. Force pushing to the PR branch has many downsides including obsoleting CI runs and making reviews anchored to commits that are no longer part of the PR.

**Comment by andreas-karlsson** [comment]
Killer feature indeed! That's currently the number one annoyance in the PR workflow. PRs getting rebased during or "after" reviews.

**Comment by connorjclark** [comment]
Please support this. I can get non experienced git devs to use fixup, but not to rebase locally. This would greatly help with PR workflows.

**Comment by Castavo** [comment]
This would be much appreciated, I have merged too many `fixup!` commits by accident :disguised_face:

**Comment by hickford** [comment]
See also GitLab feature request https://gitlab.com/gitlab-org/gitlab/-/issues/212

**Comment by justingrant** [comment]
If fixup commits aren't autosquashed on merge, at least there should be some sort of warning shown to make it harder to accidentially merge fixup commits.

**Reply by tuukkamustonen** [reply]
There are several GitHub actions for that, e.g. https://github.com/13rac1/block-fixup-merge-action

**Comment by benoitdm-oslandia** [comment]
Sorry, but I don't know how these threads work, but should we wait and hope that a GitHub core contributor drops into this discussion or is there something else that needs to be done to get an answer if this is feasible or not? (no offense here)

**Reply by dhimmel** [reply]
Not sure. I assume the more upvotes this discussion receives, the more likely it is for the request to be evaluated by someone at GitHub for inclusion in their roadmap. 99 upvotes is a good start, tell all your friends!

**Reply by alanhamlett** [reply]
I don't see any way to filter discussions by number of upvotes, so I bet GitHub employees won't notice this even with many upvotes.

**Comment by nicolalunghi-xlnx** [comment]
+1100000 here
would solve all the issue with force push

also can be used by the ```suggestion feature -> that could create fixup commits

**Reply by reitzig** [reply]
FWIW, it is already possible to manually set `fixup!` commit messages when applying suggestions, which Git will recognize just fine.

**Comment by Muzer** [comment]
100%, the lack of this feature combined with the poor ability to diff across force-pushes basically makes a multi-commit-per-PR workflow needlessly difficult.

**Comment by dubeypranav4** [comment]
yes please get this killer feature prioritised

**Comment by johannessjoberg** [comment]
Sounds like a no-brainier! 🚀

**Comment by nicola-lunghi** [comment]
any news on this?

**Comment by stempler** [comment]
Maybe helpful to others - I created a [GitHub Actions workflow](https://github.com/wetransform-os/gradle-semantic-release-version/blob/main/.github/workflows/tf-autosquash.yml) that reacts on PR comments containing `/autosquash` and then performs a rebase with autosquash. Maybe you can adapt it to your needs.

Restrictions:
- PR must be based on a branch on the same repo (not from a fork)
- User that comments must be PR author or member of the organization that owns the repo
- rebases onto current state of base branch
- CI workflows will not run on the PR after pushing the changes if using the Actions GitHub token, as workaround generate an app token with required permissions and use that

Features:
- handle special case of fixup for Renovate commits - change commit author so Renovate does not override the changes

**Reply by fardarter** [reply]
It would really be better if this was platform managed rather than token authenticated, IMO.

**Reply by stempler** [reply]
Definitely, it would be awesome to have something part of the platform, especially with the many limitations for access to secrets.
While this is not available, for us it is convenient to be able to trigger it with a comment, which works for our usual workflow.

**Reply by tspeight-optos** [reply]
I'd love a version that doesn't `rebases onto current state of base branch` - in my workflow, the branching point is useful information that shouldn't be touched.  And like everyone else, wish it were part of GitHub's merge options so that it doesn't invalidate automated checks, making me re-run CI/CD jobs with the same worktree.

**Comment by RadxaYuntian** [comment]
I talked with my teammates about our review workflow, and it would be nice to have this. Though it might interfere with commit signing.

**Comment by nodeax** [comment]
+1 this will fix the practically unusable force push functionality (orphaned comments, missing diff of older commits) in GitHub PR review view
